keps/sig-node/3759-sidecar-containers/README.md

```diff
@@ -0,0 +1,1235 @@
+  Init containers, readiness of the container will be required before starting
+  the next Init container.
+  - Pros: detaching sidecar proposal from the per-container restartPolicy
+    proposal
+  - Cons: may introduce confusion whether to inject it in the Init list or
+    regular list.
+
+
+### User Stories (Optional)
+
+User stories are covered in Motivation section.
+
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+The following is the list of hypothetical scenarios how users may decide to
+abuse the feature or use it in a non-designed way. This is an exercise to
+understand potential side effects from implementing the feature. We are looking
+for side effects like abuse of the functionality to implement something
+error-prone that sidecars were not designed for. Or causing issues like noisy
+neighbors, etc.
+
+### Scenario 1. User decides to use sidecars as a way to run regular containers
+
+At least one regular container is required by the Pod. Users may decide to run
+all other containers as sidecars and have a single container with the Sleep
+logic inside with timeout or waiting for some cancellation signal from custom
+job orchestration portal. This way users may implement execution with a time
+limit defined by that logic. This is something that users cannot express today
+“declaratively” and will be possible using new sidecar containers. One may
+imagine a scenario when a third party jobs orchestration tool converts all
+containers to sidecar containers programmatically and injects a single job
+orchestration container into the Pod. This can be seen as an abuse of the
+concept, and may lead to issues when pods terminate unexpectedly or Jobs meant
+to be run as regular containers start being run as sidecars being restarted on
+Completion. However we do not see a major problem with it
+
+Another reason for users to implement regular containers as sidecars would be to
+use any special properties kubelet may apply to sidecar containers. For example,
+if restart backoff timeouts will be minimized for sidecar containers (which was
+proposed and rejected), users may decide to use this feature for regular
+containers by running those as sidecars. Current proposal as written doesn’t
+introduce any special properties for sidecars that users may start abusing.
+
+### Scenario 2. Balloon sidecars
+
+Users may decide to start the “large request” sidecar containers early to
+pre-allocate resources for other containers. The same time asking for less
+resources for a regular containers hoping to reuse what is allocated by the
+“baloon” sidecar. This was previously impossible to pre-allocate resources like
+CPU before any Init containers run. If this pattern have any benefits, people
```

Comment:
Technically the pod cgroup is max(highest(init_container), sum(container)), so a burst init container could balloon anyway today.

Reply:
there is a resource reuse logic when scheduling regular containers that potentially can not schedule regular containers to the same place. This is especially for resource managers like CPU manager or topology manager. This balooning is locking resources early. Not a big deal though.

Reply:
added a sentence about resource managers